design-proposals/klip-47-add-source-stream-table-semantic.md

```diff
@@ -0,0 +1,122 @@
+# KLIP 47 - Add Source Stream/Table Semantic 
+
+**Author**: Boyang Chen (@boyang) | 
+**Release Target**: TBD | 
+**Status**: _In Discussion_ | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/7474
+
+**tl;dr:** _Add support for source stream/table. This will give users a read-only view of data that 
+are populated by the upstream Kafka topic, making the data ownership more clear._
+           
+## Motivation and background
+
+Users of KsqlDB have no clear ownership definition around tables or streams they created. In a “normal” database, 
+when a user creates a table the DB creates the corresponding files on disk and owns the table. 
+User may write data into those tables via `INSERT INTO` statements and/or modify/delete existing data 
+via `UPDATE` and `DELETE` statements. However, the table will be not changed by any other means. 
+
+For ksqlDB, the setup is quite differently. Specifically, user could create the database based on 
+the topics stored in Kafka as the source of truth. The defined database could be mutable by 
+using `INSERT` command to populate data into it, which will then pollute the original data source.
+Secondly, when calling `CREATE TABLE`, the actual data is neither materialized nor 
+ready for interactive queries to access. To address these gaps, we would like to add:
+
+* Syntax to add read-only stream/table as `CREATE SOURCE STREAM/TABLE` to make the data source as read-only. Any 
+insertion command will be rejected, with an example error response as:
+
+```
+Error: insertion into source stream/table is not allowed
+```   
+
+* In addition, the `CREATE SOURCE TABLE` will materialize the input source topic into a stream 
+instance which would generate a RocksDB instance for interactive queries.
+
+
+## What is in scope
+* Add `CREATE SOURCE STREAM/TABLE` syntax to KSQL
```

Comment:
Would we support CREATE OR UPDATE syntax as well? If yes maybe better to clarify too.

Reply:
Atm, we support `CREATE OR REPLACE` but it's for queries... (if `CSAS` and `CTAS`). I don't believe we support to modify (base) streams/tables created via `CS` or `CT`. Thus we would not need to support it I guess? \cc @agavra

Reply:
we do support it for base streams/tables :) it just updates the schema. you can also use `ALTER`

Reply:
What's the impact on queries that use those tables? -- How should it work for pull queries for source tables? Do we require a default value or would be just return NULL?

Reply:
> What's the impact on queries that use those tables?

There is no impact, a query could not have selected the field so it is not piped through. A `SELECT *` is resolved at query-time, so it doesn't 'magically' update when you alter a source.

> How should it work for pull queries for source tables?

Same as with any other query - if you issue the pull query after the field was added you would get it, otherwise you wouldn't.